### PR TITLE
No cython to install

### PR DESCRIPTION
**Pull request by sixpearls**
This is a start for #18 -- I believe it removes Cython as a requirement to install. I just checked, and it seems like NumPy does NOT need to be pre-installed to compile. 

I am working on including the NumPy implementation in #19, but I am not sure how to test setup without the compiler. I tried installing without NumPy or Cython pre-installed. After pip installed the NumPy requirement, it seemed to go back to compile the .c file.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,16 +1,47 @@
...
-    Extension(f"{name}._bspl",
-              [os.path.join(name, "_bspl.pyx")],
-              include_dirs=[numpy.get_include()],
-              depends=[os.path.join(name, "_bspl.h")]),
-]
+cmdclass = {}
+
+if use_cython and use_numpy:
+    extensions = cythonize([
+        Extension(f"{name}._bspl",
```
Comment (ixjlyons) [inline setup.py]: Should we make an `extension = "_bspl"` variable and if you're cool with requiring Python >= 3.6, use f strings like here to replace all occurrences of `_bspl` with `{extension}`.

Comment (ixjlyons) [reply setup.py]: A lot of projects are moving on to require > 3.4, but I think 3.5 support is still pretty widespread. For me, f strings are nice but not really worth being the sole reason for a Python version requirement.

Comment (sixpearls) [reply setup.py]: I neither feel motivated to support <= 3.5 nor do I feel strongly about f-strings over some other way to DRY up references to the name.  I guess we should be consistent though? If you don't mind I'll defer to you on this. I believe the f-string usage here came from your commit.

File: setup.py [hunk 2]
```diff
@@ -1,16 +1,48 @@
...
 
-extensions = [
-    Extension(f"{name}._bspl",
-              [os.path.join(name, "_bspl.pyx")],
-              include_dirs=[numpy.get_include()],
-              depends=[os.path.join(name, "_bspl.h")]),
-]
+    class sdist(_sdist):
+        def run(self):
+            cythonize([os.path.join(name, "{}.pyx".extension())])
```
Comment (ixjlyons) [inline setup.py]: Ok I just used format strings. I also changed this (used to be `cythonize(['cython/mycythonmodule.pyx'])`) which I think just got left over from copying the template.

Comment (ixjlyons) [reply setup.py]: Also just realized I don't like using the name `extension` and I screwed this one up. Should be fixed now

**Comment by sixpearls** [comment]
Turns out we don't need to override the sdist command at all (anymore?).

In my environment (python 3.6) with Cython installed, I can run `python setup.py sdist`, and it will cythonize the extension and generate the .c file (if it doesn't exist or (I assume) if it is stale) and add it to the `egg-info\SOURCES.txt` file. This makes me think doing an `sdist upload` to release a version to PyPI will work.

So with #19, this should be pip installable now?

Should we add the .c file to the repository so that you could install from the repository without Cython? Or say that if you are installing from the repository, you have to use Cython? Leaning towards the latter.

**Comment by ixjlyons** [comment]
:+1:` for leaving `_bspl.c` out of git

Can always upload a dry run to https://test.pypi.org/ to check that everything works.

**Comment by sixpearls** [comment]
I'm not sure why, but when I try to install from PyPI my compiler complains about not finding the header file. Could you try? Use 

``$ python -m pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple ndsplines``

to pull ndsplines from test and numpy/scipy from production

**Comment by ixjlyons** [comment]
Looks like we just need to include `ndsplines/_bspl.h` in the sdist package in `MANIFEST.in` [see here](https://github.com/pypa/packaging-problems/issues/84)

**Comment by sixpearls** [comment]
I was just looking at that. It looks like I can add the .h file to the second argument and it seems to work. I am not sure why I can't find a docs for the Extension class.

Do you mind trying again?

Uploading, after installing twine, is:

.. code:: bash
    $ python setup.py sdist
    $ python -m twine upload --repository-url https://test.pypi.org/legacy/ dist/*

**Comment by ixjlyons** [comment]
Seems to be working for me.

**Comment by sixpearls** [comment]
Okay, this worked on a computer with no compiler (although I suppose I'm not sure that the `optional` flag was necessary, I think it should have been).

Merging and closing everything.

**Pull request merged**